### ISSUE TITLE
Search with "wrap around" not working properly

### ISSUE THREAD

**Issue by xgid** [leo-editor/leo-editor#457]
I make a search of a text in the entire outline with the "**wrap Around**" option activated. I start the search in the middle of the outline, find some matches in several nodes and keep pressing F3 until I reach the end of the outline. Then the search goes on and finds some other matches at the begining of the outline and so on until the last one is found, in which moment the following message appears in the status bar:

>    not found: (ignore-case, wrapping) some_text_searched

So far, so good.

But, if at this point, I go to the end of the body of the current node (the node with the last match) and press F3 again to start a new search (without changing the text to search), the search stops right there and shows again the message:

>    not found: (ignore-case, wrapping) some_text_searched

To give more hints as to the possible cause of the problem: if, instead of going to the end of the body of the current node, I go to the begining of the body and press F3 **until all matches in this node are found**, the next F3 key press shows the error message again and the search stops there, never leaving the node.

**Comment by vitalije** [comment]
I am investigating this issue and have noticed that if there are several matches in currently selected position, find-next will find first occurrence and after that it will skip rest of the occurrences in current body and go to the next match in outline order.

**Comment by vitalije** [comment]
I believe it would be desirable if Leo after reporting there are no more matches returns focus to position where search has started.

**Comment by vitalije** [comment]
96c8d8a should fix this issue. @xgid, please reopen if it didn't solve it.

**Comment by edreamleo** [comment]
Many thanks, vitalije, for this work.  I was not looking forward to it ;-)

**Comment by xgid** [comment]
Thanks Vitalije! The original bug has been fixed, yes.

There are only two details I have seen when using it, which I don't know if are the expected/usual way that global searches work:
1. Your suggestion that "_after reporting there are no more matches returns focus to position where search has started_" is not being fulfilled. The focus remains at the last match.
1. The search does not respect the nodes I had expanded before the search: after the wrap around, only the top-level node (and subtree) where the last match is found remains expanded. All other top-level nodes are collapsed (when they originally were not).

**Comment by xgid** [comment]
My current configuration is:
```
Leo 5.5, build 20170711085521, Tue Jul 11 08:55:21 CDT 2017
Git repo info: branch = master, commit = 96c8d8ab8078
Python 3.6.1, PyQt version 5.6.2
Windows 10 AMD64 (build 10.0.15063) SP0
```

I have also just noticed this logs appearing at the Log pane after the checkout of your fix:
```
update_md markdown not available: using rst
update_md markdown not available: using rst
...
<string>:3: (WARNING/2) Inline emphasis start-string without end-string.
update_md markdown not available: using rst
```

I've been with an old master for some days, so it may have nothing to do with your changes...

**Comment by xgid** [comment]
Oh, and I have also updated my Anaconda installation to v4.4.0 (Python 3.6), so that may also be the reason...

**Comment by xgid** [comment]
Ups! I've also just been reported this Traceback in my last search:
```python
Traceback (most recent call last):
  File "C:\dev\Leo-5.4\leo\core\leoGlobals.py", line 2316, in wrapper
    func(self, event=event)
  File "C:\dev\Leo-5.4\leo\core\leoFind.py", line 434, in findNextCommand
    self.findNext()
  File "C:\dev\Leo-5.4\leo\core\leoFind.py", line 1697, in findNext
    pos, newpos = self.findNextMatch()
  File "C:\dev\Leo-5.4\leo\core\leoFind.py", line 1760, in findNextMatch
    pos, newpos = self.search()
  File "C:\dev\Leo-5.4\leo\core\leoFind.py", line 1980, in search
    if self.passedWrapPoint(p, pos, newpos):
  File "C:\dev\Leo-5.4\leo\core\leoFind.py", line 2012, in passedWrapPoint
    not self.reverse and newpos > self.wrapPos)
TypeError: '>' not supported between instances of 'int' and 'NoneType'
```

**Comment by xgid** [comment]
I was trying to reproduce it, but got it only **twice** from 6 trials! So I'm not sure how to reproduce... sorry!

**Comment by vitalije** [comment]
@xgid  thanks for your report. I'll see what I can do. My suggestion that it would be nice to end where you have started search is still a wish list item. I haven't do anything about it yet.

Concerning the expanded/collapsed part of your question, I am not sure that it is a bug. Maybe it was designed to do that, i.e. to collapse everything except what is found. Are those collapsed nodes searched properly? Or you suggest that they are not.

Maybe best thing is to try search on official leoSettings.leo. Then you can give me an example of search, where you start, what you are searching for and what you expect to happen and what actually did happen. 

Armed with such report I can (hopefully) reproduce it myself.

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 2:13 AM, Виталије Милошевић <
notifications@github.com> wrote:

> @xgid <https://github.com/xgid> thanks for your report. I'll see what I
> can do. My suggestion that it would be nice to end where you have started
> search is still a wish list item. I haven't do anything about it yet.
>
> Concerning the expanded/collapsed part of your question, I am not sure
> that it is a bug. Maybe it was designed to do that, i.e. to collapse
> everything except what is found. Are those collapsed nodes searched
> properly? Or you suggest that they are not.
>
​Note: @bool collapse_nodes_during_finds = True

Edward

**Comment by vitalije** [comment]
49c712e  solves the issue almost completely. Search command will return state as it was at the beginning of search. The only thing is that every clone gets same state (if any of them was expanded, then all of them get expanded, if all of them were contracted then they all remain as they were).
Vitalije

**Comment by edreamleo** [comment]
Commit 49c712 is probably incompatible with @bool collapse_nodes_during_finds = True

**Comment by vitalije** [comment]
No, it is not. Collapses during finds do happen. Only when search is exhausted state is restored as it was before search started.

**Comment by vitalije** [comment]
Or maybe I am misinterpreting collapse_nodes_during_finds? If set to True, does it mean that state should be all collapsed except selected position even when search has finished?

**Comment by edreamleo** [comment]
Hmm.  Maybe there needs to be a new setting.  But I think not.  And I think that restoring the expansion state of nodes at the end of a find is a dubious idea.  Why complicate matters?

Furthermore, Leo has a mechanism for expanding *positions* of cloned nodes differently. That should be respected if possible.  If you really want to explore these question, a separate issue would be appropriate.  But for now, I would recommend *not* trying to restore the original search state.  Just detect the end of the wrapped search properly ;-)

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 3:44 AM, Виталије Милошевић <
notifications@github.com> wrote:

> Or maybe I am misinterpreting collapse_nodes_during_finds? If set to True,
> does it mean that state should be all collapsed except selected position
> even when search has finished?
>

​Repost from comment left on the tracker. Apparently emails are reported
more widely:

Hmm. Maybe there needs to be a new setting. But I think not. And I think
that restoring the expansion state of nodes at the end of a find is a
dubious idea. Why complicate matters?

Furthermore, Leo has a mechanism for expanding *positions* of cloned nodes
differently. That should be respected if possible. If you really want to
explore these question, a separate issue would be appropriate. But for now,
I would recommend *not* trying to restore the original search state. Just
detect the end of the wrapped search properly ;-)

Edward

**Comment by vitalije** [comment]
@edreamleo please let me explain.

> And I think that restoring the expansion state of nodes at the end of a find is a dubious idea. 

While editing some node, user realizes that something should be changed in several places throughout the outline. Possibly those changes require some manual intervention so that using change-all is not an option. User starts search with wrap checked and Leo shows another node and selects found text. User can edit or skip that occurrence, presses F3 and search continues. When there is no more hits, user would be left in some node who knows where from the start of search. Now if user wants to continue work where it was interrupted by search, he/she needs to manually find the node where search has started. Wouldn't it be nice if Leo did restore selected position and outline just as it were before search had been started?

> Why complicate matters?

Well, I have already done this item. Do you want me to delete the code I wrote? Or is it O.K. with you if I add setting for this not_found_returns_to_start_position with default value False, and if it is not set, disable new code.

OP and myself have found this feature desirable.
Vitalije

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 4:25 AM, Виталије Милошевић <
notifications@github.com> wrote:

When there is no more hits, user would be left in some node who knows where
> from the start of search. Now if user wants to continue work where it was
> interrupted by search, he/she needs to manually find the node where search
> has started. Wouldn't it be nice if Leo did restore selected position and
> outline just as it were before search had been started?


​A good description of what we are discussing.​


> Why complicate matters?
>
> Well, I have already done this item. Do you want me to delete the code I
> wrote? Or is it O.K. with you if I add setting for this
> not_found_returns_to_start_position with default value False, and if it
> is not set, disable new code.
>
> OP and myself have found this feature desirable.
>
​I'm convinced, for now.  Let's leave the code as it is and see how it
works in practice.

I do not want to add yet another obscure ​setting relating to searches.  I
think I can justify your choice as follows: the question about whether to
restore the previous position only arises when doing a wrapped search.  In
that case (and *only *in that case), there is well-defined starting point
for the search. So it seems reasonable restore that point when the search
ends. Furthermore, the "restoration" shouldn't be too jarring for the user,
because Leo will report "end of wrapped search".

In short.  Good work Vitalije.  I'll close the item.

Edward

**Comment by edreamleo** [comment]
The present code is thoroughly broken for non-wrapped searches.

It is NOT permissible to change c.p when a non-wrapped search fails.  Please fix this immediately.

Also, 72a9b5 fixed a python 3 crasher.  I suggest always developing with python 3.  That way you will avoid more compatibility issues than if you use python 2.

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 4:44 AM, Edward K. Ream <edreamleo@gmail.com> wrote:

In short.  Good work Vitalije.  I'll close the item.
>

​I am about to push a fix so that non-wrapped searches work as before.
Please don't change leoFind.py until I do.

Edward

**Comment by vitalije** [comment]
It is now disabled for non wrap searches.
But If I may suggest that it is something I would expect from search to return me to place where I started search regardless of wrap checkbox. If user starts search that has not even one hit, Leo won't change currently selected position. Whenever user executes any of search commands he/she is prepared that c.p will be changed. That should n

**Comment by vitalije** [comment]
... That should not surprise user. Whenever new search is started (search text is changed or any of its check-boxes), new start position would be remembered.

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 5:56 AM, Edward K. Ream <edreamleo@gmail.com> wrote:

>
>
>
>
> ​I am about to push a fix so that non-wrapped searches work as before.
> Please don't change leoFind.py until I do.
>

​Rev d9b1e56 contains what I have called an emergency fix.  It works for me
and all tests pass. Let me know if there are problems.

Edward

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 5:56 AM, Виталије Милошевић <
notifications@github.com> wrote:

> It is now disabled for non wrap searches.
>
​I haven't seen those changes when I pulled, which is good because
otherwise they would have created conflicts.​

> But If I may suggest that it is something I would expect from search to
> return me to place where I started search regardless of wrap checkbox.
>
​I would hate that mode of operation.  This is what I meant when I said the
code was driving me crazy.

But we have a rule never to argue about preferences.  Feel free to add a
new setting. Make sure to base the code on d9b1e56.

Edward​

**Comment by vitalije** [comment]
O.k.

**Comment by edreamleo** [comment]
Meh, it seems I may have broken the wrap-around mode.

**Comment by vitalije** [comment]
I'll see to restore it. Should not be hard.

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 6:12 AM, Виталије Милошевић <
notifications@github.com> wrote:

> O.k.
>
​It looks like I broke your new code.  My apologies.

I did say that the changes were an emergency fix.  Perhaps I should have
used self.wrap instead of self.wrapping.  Let me see if I can make
everything work as we both want.

Edward

**Comment by vitalije** [comment]
I will redirect  discussion regarding preferences in leo-editor group to see if there is any interest in having such setting. Personally I don't use search commands but quicksearch instead.

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 6:19 AM, Edward K. Ream <edreamleo@gmail.com> wrote:

​> ​
Let me see if I can make everything work as we both want.

​Done at bec5e26. This issue is now closed.

Please reopen if there are further problems.

Edward

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 6:20 AM, Виталије Милошевић <
notifications@github.com> wrote:

> I will redirect discussion regarding preferences in leo-editor group to
> see if there is any interest in having such setting.
>
​Good. Thank you.

Edward
